crates/sui-indexer-alt-graphql/src/api/types/object.rs

```diff
@@ -227,27 +237,42 @@ impl Object {
             return Ok(None);
         };
 
-        Object::from_version(stored)
+        Object::from_version(scope, stored)
     }
 
     /// Load the object at the given ID and version from the store, and return it fully inflated
     /// (with contents already fetched). Returns `None` if the object does not exist (either never
-    /// existed or was pruned from the store).
+    /// existed, was pruned from the store, or did not exist at the checkpoint being viewed).
     pub(crate) async fn at_version(
         ctx: &Context<'_>,
+        scope: Scope,
         address: SuiAddress,
         version: UInt53,
     ) -> Result<Option<Self>, RpcError<Error>> {
-        let Some(c) = contents(ctx, address, version).await? else {
+        let pg_loader: &Arc<DataLoader<PgReader>> = ctx.data()?;
+
+        let contents = contents(ctx, address, version);
+        let stored_version =
+            pg_loader.load_one(VersionedObjectVersionKey(address.into(), version.into()));
+        let (contents, stored_version) = join!(contents, stored_version);
+
+        let Some(c) = contents? else {
             return Ok(None);
         };
 
-        Ok(Some(Self::from_contents(c)))
+        if stored_version
+            .context("Failed to get object version")?
+            .is_none_or(|s| s.cp_sequence_number as u64 > scope.checkpoint_viewed_at())
+        {
+            return Ok(None);
+        }
```

Comment:
To support consistency, we need to check which checkpoint an object update happened in, and discard it if it happened ahead of the checkpoint being viewed (otherwise we risk nested queries trying to fetch from tables that aren't ready yet).

@phoenix-o / @bmwill, would it be feasible to include the checkpoint that an object update was outputted by in the columns returned by the archival service?

It's not urgent, so this can wait for the proto backfill, but it would help avoid having to make a request to the DB and to the KV store to fetch an object at a specific version.

Reply:
We don't store this information in the object.
But you can utilize the [previous_transaction field](https://github.com/MystenLabs/sui/blob/1dabb5c2df9ac98ded07fbb00eb8620d6e619c6b/crates/sui-types/src/object.rs#L643) in the object to derive the checkpoint.
Of course, this will add some extra latency due to the additional lookup

Reply:
Yeah that would be a regression on what we're doing today (which is to send a request to the `obj_versions` table in postgres at the same time as we send the request to the KV store for the contents), the question is whether we can add this as a column to the KV store, even though it is not part of the on-chain representation of an object, because it should always be possible to determine it at the time the object is being indexed (as objects are being indexed checkpoint-wise).

Reply:
do you need a backfill for this? or it's mostly concerns the latest data so we can keep it Optional?

Reply:
I think for it to be useful we would need a backfill, but it is also not urgent -- the way we're doing it here by sending the two concurrent request is fine for now, it just means that it's difficult to run GraphQL without Postgres (just talking to KV store, and/or just talking to fullnode) -- it would be fine to wait and handle this with the backfill that changes the format to proto (this change may also have implications to the proto format for objects in general?)